### PR TITLE
feat(boolean-validation): Added validation functions and test cases for Boolean data type

### PR DESCRIPTION
**Pull request by shraddha1704**
### What
* Added `truthy` and `falsy` validation functions for Boolean data type.
* Convert to `boolean` type

### Why
* User can now define schema for Boolean data type and specify values that should be treated as `true` or `false`.

### CODE REVIEW

File: lib/type/boolean.ex [hunk 1]
```diff
@@ -1,15 +1,52 @@
...
+  @spec truthy_falsy_validate(t, binary, map) :: {:ok, map} | {:error, binary}
+  defp truthy_falsy_validate(%__MODULE__{falsy: falsy, truthy: truthy}, field, params) do
+    cond do
+      !Map.has_key?(params, field) ->
+        {:ok, params}
+
+      params[field] in Enum.uniq(truthy ++ @truthy_default) ->
+        {:ok, Map.replace!(params, field, true)}
+
+      params[field] in Enum.uniq(falsy ++ @falsy_default) ->
```
Comment (jhwang09) [inline lib/type/boolean.ex]: this logic seems like is duplicate, we can abstract to a separate function

File: lib/type/boolean.ex [hunk 2]
```diff
@@ -1,15 +1,52 @@
...
+    end
+  end
+
+  @spec truthy_falsy_validate(t, binary, map) :: {:ok, map} | {:error, binary}
+  defp truthy_falsy_validate(%__MODULE__{falsy: falsy, truthy: truthy}, field, params) do
+    cond do
+      !Map.has_key?(params, field) ->
+        {:ok, params}
+
+      params[field] in Enum.uniq(truthy ++ @truthy_default) ->
```
Comment (jhwang09) [inline lib/type/boolean.ex]: do we need to check case sensitivity on the truthy value? e.g. if truthy is ["true"], do we need to account for inputs like "tRue"?

**Review by jhwang09 — COMMENTED** [review]
some commnets

File: lib/type/boolean.ex (hunk 2 above)
Comment (shraddha1704) [reply lib/type/boolean.ex]: I hadn't considered that case. Joi supports it. I think we should too. What does everyone think? @lob/atlas

File: lib/type/boolean.ex [hunk 3]
```diff
@@ -1,15 +1,79 @@
...
+  def validate_field(type, field, data) do
+    with {:ok, data} <- Required.validate(type, field, data),
+         {:ok, data} <- truthy_falsy_validate(type, field, data) do
+      {:ok, data}
+    else
+      {:error, msg} -> {:error, msg}
+    end
+  end
+
+  @spec check_boolean_values(term, [term], [term]) :: {:ok, map} | nil
```
Comment (kylekwong) [inline lib/type/boolean.ex]: i think this should be `:: true | nil`

File: lib/type/boolean.ex [hunk 4]
```diff
@@ -1,15 +1,79 @@
...
+  def check_boolean_values(initial_value, additional_values, default_values)
+      when is_binary(initial_value) do
+    allowed_values = Enum.uniq(additional_values ++ default_values)
```
Comment (kylekwong) [inline lib/type/boolean.ex]: How about

```elixir
allowed_values =
  additional_values
  |> &(&1 ++ default_values).()
  |> Enum.uniq()
  |> Enum.map(fn item -> ...) 
```

File: lib/type/boolean.ex [hunk 5]
```diff
@@ -1,15 +1,79 @@
...
+
+    allowed_values =
+      Enum.map(allowed_values, fn item ->
+        if is_binary(item) do
+          String.downcase(item)
+        end
+      end)
+
+    if String.downcase(initial_value) in allowed_values do
```
Comment (kylekwong) [inline lib/type/boolean.ex]: You can make this one line:

```elixir
if String.downcase(initial_value) in allowed_values, do: true
```

By default, functions return nil, so if you leave out the `else` case the behavior will be the same.

File: lib/type/boolean.ex [hunk 6]
```diff
@@ -1,15 +1,79 @@
...
+      true
+    else
+      nil
+    end
+  end
+
+  def check_boolean_values(initial_value, additional_values, default_values) do
+    if initial_value in Enum.uniq(additional_values ++ default_values) do
```
Comment (kylekwong) [inline lib/type/boolean.ex]:
```elixir
if initial_value in Enum.uniq(additional_values ++ default_values), do: true
```

File: lib/type/boolean.ex [hunk 7]
```diff
@@ -1,15 +1,79 @@
...
+    end
+  end
+
+  @spec truthy_falsy_validate(t, binary, map) :: {:ok, map} | {:error, binary}
+  defp truthy_falsy_validate(%__MODULE__{falsy: falsy, truthy: truthy}, field, params) do
+    cond do
+      !Map.has_key?(params, field) ->
+        {:ok, params}
+
+      check_boolean_values(params[field], truthy, @truthy_default) ->
```
Comment (kylekwong) [inline lib/type/boolean.ex:55]: Is it possible to have both truthy and falsy defined? In that case, if this evaluates to true, doesn't that mean line 65 will never be run?

**Review by kylekwong — CHANGES_REQUESTED** [review]
Some questions/suggestions

File: lib/type/boolean.ex (hunk 7 above)
Comment (shraddha1704) [reply lib/type/boolean.ex:55]: Discussed in person. It's not possible

**Review by kylekwong — APPROVED** [review]

File: lib/type/boolean.ex [hunk 8]
```diff
@@ -1,15 +1,72 @@
...
+  def validate_field(type, field, data) do
+    with {:ok, data} <- Required.validate(type, field, data),
+         {:ok, data} <- truthy_falsy_validate(type, field, data) do
+      {:ok, data}
+    else
+      {:error, msg} -> {:error, msg}
+    end
+  end
+
+  @spec check_boolean_values(term, [term], [term]) :: true | nil
```
Comment (jhwang09) [inline lib/type/boolean.ex]: for the return values, can they be `true` or `false`? that way you can update the signature to boolean type

File: lib/type/boolean.ex [hunk 9]
```diff
@@ -1,15 +1,72 @@
...
+      additional_values
+      |> (&(&1 ++ default_values)).()
+      |> Enum.uniq()
+      |> Enum.map(fn item ->
+        if is_binary(item) do
+          String.downcase(item)
+        end
+      end)
+
+    if String.downcase(initial_value) in allowed_values, do: true
```
Comment (jhwang09) [inline lib/type/boolean.ex]: per previous comment, you can add a else statement here

File: lib/type/boolean.ex (hunk 8 above)
Comment (jhwang09) [reply lib/type/boolean.ex]: also, should this function be private? I see that the private function `truthy_falsy_validate` is using this, but I don't see other module using it

**Review by jhwang09 — APPROVED** [review]

**Pull request merged**